### PR TITLE
#24 pre-cache enviroment prior to verifyWithHCaptcha call

### PR DESCRIPTION
**Pull request by CAMOBAP**
Closing https://github.com/hCaptcha/hcaptcha-android-sdk/issues/24 
 
 - Fragment instantiation now happens on `new HCaptcha`
 - WebView to use cache before reaching network
 - Update sample application

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,7 +92,7 @@ final HCaptchaConfig config = HCaptchaConfig.builder()
                 .resetOnTimeout(true)
                 .theme(HCaptchaTheme.DARK)
                 .build();
-HCaptcha.getClient(this).verifyWithHCaptcha(config)...;
+HCaptcha.getClient(this, config).verifyWithHCaptcha()...;
```
Comment (DSergiu) [inline README.md:95]: This would be a breaking change. 

Can we somehow maintain the old api and add a new parameter to the `HCaptchaConfig.lazy = true/false` which would basically just not `fragment.show()` but require an explicit `.show()` to be called to show the hCaptcha view. Thus, the new api is not a breaking change and allows for both old and new lazy approach.

 What do you reckon @CAMOBAP @e271828- ?

Example:

```java
final HCaptchaConfig config = HCaptchaConfig.builder()
                ...
                .theme(HCaptchaTheme.DARK)
                .lazy(true)
                .build();
final HCaptcha hCaptchaClient = HCaptcha.getClient(this).verifyWithHCaptcha(config);

// Async when ready to show, client would do:
hCaptchaClient.show()
  .addOnSuccessListener(new OnSuccessListener<HCaptchaTokenResponse>() {
        @Override
        public void onSuccess(HCaptchaTokenResponse response) {
            String userResponseToken = response.getTokenResult();
            // Validate the user response token using the hCAPTCHA siteverify API.
        }
    })
    .addOnFailureListener(new OnFailureListener() {
        @Override
        public void onFailure(HCaptchaException e) {
            Log.d("MainActivity", "Error code: " + e.getStatusCode());
            Log.d("MainActivity", "Error msg: " + e.getMessage());
        }
    });
```

File: example-app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -11,8 +11,14 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
+
+        <meta-data
+            android:name="com.hcaptcha.sdk.site-key"
+            android:value="10000000-ffff-ffff-ffff-000000000001" />
```
Comment (DSergiu) [inline example-app/src/main/AndroidManifest.xml:17]: Was this a new requirement to allow the sitekey via manifest meta data? I do not see it being part of Issue #24?

File: README.md (hunk 1 above)
Comment (e271828-) [reply README.md:95]: This seems like a better approach, yeah.

Comment (CAMOBAP) [reply README.md:95]: > This would be a breaking change.

Definitely

> What do you reckon @CAMOBAP @e271828- ?

In this case, API becomes less clear for me because depends on `lazy` param `verifyWithHCaptcha` will work significantly different, + extra API method `show`

So between `less clear API` vs `breaking change` I would select the first one

File: example-app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (DSergiu) [reply example-app/src/main/AndroidManifest.xml:17]: This looks like a nice feature indeed, but it can be confusing to users using multiple sitekeys. Can this be moved to a separate PR, or maybe open a ticket for approval first? It would be nice to keep the PR focused on the ticket it is solving.

Comment (CAMOBAP) [reply example-app/src/main/AndroidManifest.xml:17]: It can be but to be on the same page this feature is completely optional.

SDK will check `meta-data` only in case if user didn't passed it explicitly (if user called `getClient(context)`), so it's not break any existing behavior.

Ticket https://github.com/hCaptcha/hcaptcha-android-sdk/issues/26

Comment (DSergiu) [reply example-app/src/main/AndroidManifest.xml:17]: Understood. It just now gives more options to configure the sitekey. We need to make sure that the siteykey passed via `HCaptchaConfig` has higher precedence over the meta data.

Comment (CAMOBAP) [reply example-app/src/main/AndroidManifest.xml:17]: For sure, this is how it's done right now but I will add a dedicated test for this also

File: README.md (hunk 1 above)
Comment (CAMOBAP) [reply README.md:95]: Looks great, will update PR soon

Comment (CAMOBAP) [reply README.md]: Done

File: README.md [hunk 3]
```diff
@@ -60,7 +60,28 @@ HCaptcha.getClient(this).verifyWithHCaptcha(YOUR_API_SITE_KEY)
...
+</manifest>
+```
+
+And then simply call:
+
+```java
+HCaptcha.getClient(this).verifyWithHCaptcha()...
+```
+
+If you need to customize the look and feel, language, endpoint, etc. this can be achieved by passing a `HCaptchaConfig` object to `getClient` method.
```
Comment (DSergiu) [inline README.md:84]: This is outdated.

File: example-app/src/main/res/layout/activity_main.xml [hunk 4]
```diff
@@ -7,34 +7,59 @@
...
+    <RadioGroup
+        android:id="@+id/sizeRadioGroup"
         android:layout_width="match_parent"
-        android:layout_height="wrap_content">
-
-        <Button
+        android:layout_height="wrap_content"
+        android:orientation="horizontal">
+        <RadioButton
+            android:id="@+id/size_invisible"
```
Comment (DSergiu) [inline example-app/src/main/res/layout/activity_main.xml:16]: I like the new configuration buttons. I do have this issue where the text is not visible.

Device: 
image


image

File: example-app/src/main/java/com/hcaptcha/example/MainActivity.java [hunk 5]
```diff
@@ -14,22 +21,65 @@
...
+        hCaptchaConfig = builder.build();
+
+        hCaptcha = HCaptcha.getClient(this);
+        if (setupCheckBox.isChecked()) {
+            hCaptcha.setup(hCaptchaConfig);
+        }
     }
 
     private void setTokenTextView(final String text) {
-        tokenTextView.setText(text.substring(0, Math.min(100, text.length())) + "...");
```
Comment (DSergiu) [inline example-app/src/main/java/com/hcaptcha/example/MainActivity.java:32]: This change shows the whole token, but there is no scrolling, thus it looks odd. Why the change?

image

Comment (CAMOBAP) [reply example-app/src/main/java/com/hcaptcha/example/MainActivity.java:32]: Fixed

File: README.md (hunk 3 above)
Comment (CAMOBAP) [reply README.md]: Done

File: example-app/src/main/res/layout/activity_main.xml (hunk 4 above)
Comment (CAMOBAP) [reply example-app/src/main/res/layout/activity_main.xml:17]: Fixed

File: README.md [hunk 6]
```diff
@@ -74,6 +95,16 @@ final HCaptchaConfig config = HCaptchaConfig.builder()
...
 
+Also to reduce hCaptcha start time, `setup` can be used:
+
+```java
+HCaptcha hCaptcha = HCaptcha.getClient(this).setup()
+```
+
+`setup` accepts the same arguments as `verifyWithHCaptcha`. Once `setup` is called, later you can call `verifyWithHCaptcha` with no arguments
+
+If `verifyWithHCaptcha` will be called with different arguments than `setup`, SDK will handle this and re-configure hCaptcha.
```
Comment (e271828-) [inline README.md:106]:
```suggestion
If `verifyWithHCaptcha` is called with different arguments than `setup` the SDK will handle this by re-configuring hCaptcha.
```

Comment (CAMOBAP) [reply README.md:106]: I would say that with a new config object it will "work as before", so no benefits, because it will create a fragment instance.

We added "load from the cache" to WebView, but it doesn't look related to this place in the documentation

File: CHANGES.md [hunk 7]
```diff
@@ -1,4 +1,8 @@
 # Changelog
 
+## 2.1.0
+
+- Add `HCaptcha.setup` method ([#24](https://github.com/hCaptcha/hcaptcha-android-sdk/issues/24))
```
Comment (e271828-) [inline CHANGES.md:5]:
```suggestion
- Add `HCaptcha.setup` method to improve cold-start time, enable asset caching ([#24](https://github.com/hCaptcha/hcaptcha-android-sdk/issues/24))
```

**Review by e271828- — APPROVED** [review]
